base/compiler/typeinfer.jl

```diff
@@ -8,7 +8,179 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
     return typeinf(interp, frame)
 end
 
+"""
+The module `Core.Compiler.Timings` provides a simple implementation of nested timers that
+can be used to measure the exclusive time spent inferring each method instance that is
+recursively inferred during type inference.
+
+This is meant to be internal to the compiler, and makes some specific assumptions about
+being used for this purpose alone.
+"""
+module Timings
+
+using Core.Compiler: -, +, :, length, push!, pop!, @inline, @inbounds, copy
+
+# What we record for any given frame we infer during type inference.
+function _typeinf_identifier(frame::Core.Compiler.InferenceState)
+    # TODO: What information do we want to collect to identify each invocation of type inference?
+    mi_info = (
+        frame.linfo,
+        frame.world,
+        copy(frame.sptypes),
+        frame.slottypes[2:end],  # Skip repeating the method instance
+    )
+    return mi_info
+end
+
+"""
+    Core.Compiler.Timing(mi_info, start_time, ...)
+
+Internal type containing the timing result for running type inference on a single
+MethodInstance.
+"""
+struct Timing
+    mi_info::Any
+    start_time::UInt64
+    cur_start_time::UInt64
+    time::UInt64
+    children::Core.Array{Timing,1}
+    Timing(mi_info, start_time, cur_start_time, time, children) = new(mi_info, start_time, cur_start_time, time, children)
+    Timing(mi_info, start_time) = new(mi_info, start_time, start_time, UInt64(0), Timing[])
+end
+
+_time_ns() = ccall(:jl_hrtime, UInt64, ())  # Re-implemented here because Base not yet available.
+
+# We keep a stack of the Timings for each of the MethodInstances currently being timed.
+# Since type inference currently operates via a depth-first search (during abstract
+# evaluation), this vector operates like a call stack. The last node in _timings is the
+# node currently being inferred, and its parent is directly before it, etc.
+# Each Timing also contains its own vector for all of its children, so that the tree
+# call structure through type inference is recorded. (It's recorded as a tree, not a graph,
+# because we create a new node for duplicates.)
+const _timings = Timing[]
+"""
+    Core.Compiler.reset_timings()
+Empty out the previously recorded type inference timings, and start the "root" timer again.
+"""
+function reset_timings()
+    Core.Compiler.empty!(_timings)
+    Core.Compiler.push!(_timings, Timing("root", _time_ns()))
```

Comment:
Oh! Wait, one thing i'd like to fix before merging is to make the root node have a consistent type as all the other nodes.

Currently the other nodes' `mi_info`s are `Tuple{Core.MethodInstance,UInt64,Vector{Any},Vector{Any}}`.

I had them as an `::Any` because i was still hacking on it. lemme change it to this, but then i'm not sure what to put for the MethodInstance for `"root"`....

Maybe I could make a dummy method instance for a dummy "root" function or something? Any other suggestion?

Reply:
Do you really need a dummy root? `close_current_timer` and `exit_current_timer` could check to see if the stack is empty. Or do they need to record something in some parent somewhere? If you really do need a dummy node, perhaps make it a `Union{Nothing, Tuple{...}}` and just set it to `nothing`?

Reply:
Hrmmmmm, maybe you're right! Good suggestion! I think it was mainly there to make the algorithm simpler, but you're right we probably don't need it!

It's currently storing the complement of the inference time, which is interesting philosophically, but I think we can reconstruct that in post-processing. Great suggestion, i'll delete this today. :)